### ISSUE TITLE
Join a group

### ISSUE THREAD

**Issue by evanp** [swicg/groups#1]
"As an ActivityPub user, I want to join a group, so I can be part of it."

**Comment by mayel** [comment]
Many implementations currently use `Follow`/`Undo(Follow)` for group membership ([FEP-1b12](https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md) pattern).

However, this conflates two distinct concepts:
1. **Following** (subscription) - "Show me this group's posts in my feed"
2. **Membership** (permissions) - "I'm part of this community with certain rights"

Use cases this creates problems for:
- A user wants to follow a public group's posts without formally joining
- A member wants to stay in the group but "silence" it (unfollow without leaving)
- Different permission levels for followers vs actual members

A backwards-compatible solution could be to support both patterns:
- `Follow` / `Accept` / `Undo(Follow)` for subscription (receiving posts)
- `Join` / `Accept` / `Leave` for membership (having permissions) for implementations that suport it
- Progressive enhancement: for open groups `Follow` can also make you a member, and for closed groups mods can add/approve followers as members, which makes following equivalent to a "request to join"
- Expose separate `followers` and `members` collections on Group actors

Example group actor:
```json
{
  "type": "Group",
  "name": "A sample group",
  "followers": "https://example.org/groups/1/followers",
  "members": "https://example.org/groups/1/members"
}
```

**Comment by trwnh** [comment]
> Many implementations currently use `Follow`/`Undo(Follow)` for group membership ([FEP-1b12](https://codeberg.org/fediverse/fep/src/branch/main/fep/1b12/fep-1b12.md) pattern).
> 
> However, this conflates two distinct concepts:
> 
> 1. **Following** (subscription) - "Show me this group's posts in my feed"
> 
> 2. **Membership** (permissions) - "I'm part of this community with certain rights"

Agreed. It should be possible to both Follow and Join a Group. Follow adds you to the as:followers collection, and Join adds you to the _:members collection. (loosely, anyway -- the exact data structures and vocab terms are subject to change.)

We also want to avoid conflating the Group concept where "the members comprising the group" is distinct from "the construct managing membership".

> A backwards-compatible solution could be to support both patterns:
> 
> * `Follow`/`Undo(Follow)` for subscription (receiving posts)
> 
> * `Join`/`Leave` for membership (having permissions) for implementations that suport it

When you put it like this, it does seem weird that we either don't have an Unfollow activity, or that we use Leave instead of Undo(Join). What do we do about this?

**Comment by trwnh** [comment]
I'm kind of wondering whether it would make sense to treat this as a subset case of #4 -- in other words, to join a group, you need an invite, but there can be a public invite that lets anyone join. Then *after* you become a member, a Join activity is generated to indicate that you successfully joined the group.

Approval as per #10 could be combined with this.

Alternatively, some kind of invite token would be referenced via a Join activity (`instrument`?) and you would still need to depend on an Accept Join arriving later. But I think it would be better if the Join activity is controlled by the group[^1], separately from a "join request" or "using an invite". Looking at other systems, we could use "membership records" which get created by the user who wants to join, default to pending/inactive, and later get updated to no longer be pending (aka active).

[^1]: We have a similar problem with Follow activities right now where there is no in-spec way to remove a follower that has already been accepted, and it's problematic to have to notify followers that they've been removed and/or having to synchronize follower state somehow in cases of shared inbox delivery to followers collections, instead of simply stopping deliveries to them. Having the state managed on the delivering/controlling side avoids this issue.

So maybe something like this flow:

```
id: <GrcUeeCwD2>
type: _:InvitationToken
expireAfter: 2026-04-24T00:00:00Z
remainingUses: 4
forGroup: <group>
grantRoles: <member>
```

```
actor: <alice>
type: _:JoinRequest
object: <group>
audience: <group>
instrument: <GrcUeeCwD2>
```

```
actor: <group>
type: Add
object: <alice>
target: <member>
```

```
actor: <alice>
type: Join
object: <group>
```

Or using membership records:

```
GET /GrcUeeCwD2

id: <GrcUeeCwD2>
type: _:InvitationToken
expireAfter: 2026-04-24T00:00:00Z
remainingUses: 4
forGroup: <group>
grantRoles: <member>
```

```
POST /members/

type: _:MembershipRecord
group: <group>
user: <alice>
invite: <GrcUeeCwD2>
pending: true
roles: ...
permissions: ...

201 Created
Location: /members/alice
```

```
PUT /members/alice

type: _:MembershipRecord
group: <group>
user: <alice>
invite: <GrcUeeCwD2>
pending: false
roles: <member>
permissions: ...
```

We could also take inspiration from MLS groups and their proposal-then-commit system:

```
type: _:Proposal
object:
  type: Add
  object: <alice>
  target: <group>
```

```
type: _:Commit
object:
  type: Add
  object: <alice>
  target: <group>
```

**Comment by evanp** [comment]
So, I think we should treat FEP 1b12 group compatibility as orthogonal.

Let's get the `Join` + `Leave` structure down first, and then worry about `Follow` + `Undo`/`Follow` later.

**Comment by evanp** [comment]
To join a group ([Join a group][1]), an actor sends a `Join` activity to the group. The activity must include the `object` property, which is the group to join.

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "https://example.com/activities/1234567890",
  "actor": "https://example.com/actors/3456789012",
  "type": "Join",
  "to": {
    "type": "Group",
    "id": "https://example.com/groups/world-group",
    "name": "World Group"
  },
  "object": {
    "id": "https://example.com/groups/world-group",
    "type": "Group",
    "name": "World Group"
  }
}
```

**Comment by evanp** [comment]
🚪 **Alternative** The above design requires the `Group` to be an actor that can receive activities. An alternative design would have the actor send the join request to the group's owner and/or admins.

```json
{
  "@context": "https://www.w3.org/ns/activitystreams",
  "id": "https://example.com/activities/1234567890",
  "actor": "https://example.com/actors/3456789012",
  "type": "Join",
  "object": {
    "id": "https://example.com/groups/world-group",
    "type": "Group",
    "name": "World Group"
  },
  "to": [
    {
      "id": "https://example.com/actors/3456789012",
      "type": "Person",
      "name": "Evan Prodromou"
    },
    {
      "id": "https://example.com/groups/world-group/admins",
      "type": "OrderedCollection",
      "name": "World Group admins"
    }
  ]
}
```
